// @tago-io/sdk/out/modules/Connector/Connector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Connector extends TagoIOModule_1.default {
    /**
     * Get information about the current connector
     */
    async info() {
        const result = await this.doRequest({
            path: "/info",
            method: "GET",
            params: {
                details: this.params.details,
            },
        });
        return result;
    }
    /**
     * Get a valid token using token serie
     * @param serieNumber
     * @param authorization
     */
    async resolveToken(serieNumber, authorization) {
        let path = `/connector/resolve/${serieNumber}`;
        if (authorization)
            path = `${path}/${authorization}`;
        const result = await this.doRequest({
            path,
            method: "GET",
            params: {
                details: this.params.details,
            },
        });
        return result;
    }
}
exports.default = Connector;
