// @tago-io/sdk/out/modules/Services/Services.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
const Console_1 = __importDefault(require("./Console"));
const SMS_1 = __importDefault(require("./SMS"));
const Email_1 = __importDefault(require("./Email"));
const MQTT_1 = __importDefault(require("./MQTT"));
const Notification_1 = __importDefault(require("./Notification"));
const Attachment_1 = __importDefault(require("./Attachment"));
class Services extends TagoIOModule_1.default {
    constructor() {
        super(...arguments);
        this.console = new Console_1.default(this.params);
        this.sms = new SMS_1.default(this.params);
        this.email = new Email_1.default(this.params);
        this.MQTT = new MQTT_1.default(this.params);
        this.Notification = new Notification_1.default(this.params);
        this.Attachment = new Attachment_1.default(this.params);
    }
}
exports.default = Services;
