// @tago-io/sdk/out/modules/RunUser/SDB.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class SDB extends TagoIOModule_1.default {
    /**
     * Retrieves a custom parameter of a Run user.
     * The Run user is identified by the token in the constructor.
     * @param tagoRunURL TagoIO Run url without http
     * @param key Identifier of the parameter
     */
    async getItem(tagoRunURL, key) {
        const result = await this.doRequest({
            path: `/run/${tagoRunURL}/sdb/${key}`,
            method: "GET",
        });
        return result;
    }
    /**
     * Creates or updates a custom parameter of a Run user.
     * The Run user is identified by the token in the constructor.
     * @param tagoRunURL TagoIO Run url without http
     * @param key Identifier of the parameter
     * @param value Value of the parameter
     */
    async setItem(tagoRunURL, key, value) {
        const result = await this.doRequest({
            path: `/run/${tagoRunURL}/sdb/${key}`,
            method: "POST",
            body: value,
        });
        return result;
    }
    /**
     * Delete a custom parameter of a Run user.
     * The Run user is identified by the token in the constructor.
     * @param tagoRunURL TagoIO Run url without http
     * @param key Identifier of the parameter
     */
    async removeItem(tagoRunURL, key) {
        const result = await this.doRequest({
            path: `/run/${tagoRunURL}/sdb/${key}`,
            method: "DELETE",
        });
        return result;
    }
}
exports.default = SDB;
