// @tago-io/sdk/out/modules/Account/Buckets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Buckets extends TagoIOModule_1.default {
    /**
     * Retrieves a list with all buckets from account
     * @example
     * Default: {
     *   page: 1,
     *   fields: ["id", "name"],
     *   filter: {},
     *   amount: 20,
     *   orderBy: "name,asc",
     * }
     * @param query Search query params
     */
    list(query) {
        const result = this.doRequest({
            path: "/bucket",
            method: "GET",
            params: {
                page: (query === null || query === void 0 ? void 0 : query.page) || 1,
                fields: (query === null || query === void 0 ? void 0 : query.fields) || ["id", "name"],
                filter: (query === null || query === void 0 ? void 0 : query.filter) || {},
                amount: (query === null || query === void 0 ? void 0 : query.amount) || 20,
                orderBy: (query === null || query === void 0 ? void 0 : query.orderBy) ? `${query.orderBy[0]},${query.orderBy[1]}` : "name,asc",
            },
        });
        return result;
    }
    /**
     * Generates and retrieves a new bucket for the account
     * @param createParams Bucket create object
     */
    create(createParams) {
        const result = this.doRequest({
            path: "/bucket",
            method: "POST",
            body: createParams,
        });
        return result;
    }
    /**
     * Modifies any property of the bucket.
     * @param bucketID Bucket ID
     * @param bucketObject JSON of bucket to replace
     */
    edit(bucketID, bucketObject) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}`,
            method: "PUT",
            body: bucketObject,
        });
        return result;
    }
    /**
     * Deletes a bucket from the account
     * @param bucketID Bucket ID
     */
    delete(bucketID) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}`,
            method: "DELETE",
        });
        return result;
    }
    /**
     * Gets information about the bucket
     * @param bucketID Bucket ID
     */
    info(bucketID) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}`,
            method: "GET",
        });
        return result;
    }
    /**
     * Get Amount of data on the Bucket
     * @param bucketID Bucket ID
     */
    amount(bucketID) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}/data_amount`,
            method: "GET",
        });
        return result;
    }
    /**
     * List variables inside the bucket
     * @example
     * Options Default: {
     *   showAmount: false
     *   showDeleted: false
     *   resolveOriginName: false
     * }
     * @param bucketID Bucket ID
     * @param options Request options
     */
    listVariables(bucketID, options) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}/variable`,
            method: "GET",
            params: {
                amount: (options === null || options === void 0 ? void 0 : options.showAmount) || false,
                deleted: (options === null || options === void 0 ? void 0 : options.showDeleted) || false,
                resolveOriginName: (options === null || options === void 0 ? void 0 : options.resolveOriginName) || false,
            },
        });
        return result;
    }
    /**
     * Delete a bucket variable
     * @param bucketID Bucket ID
     * @param deleteParams Variable Details
     */
    deleteVariable(bucketID, deleteParams) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}/variable`,
            method: "DELETE",
            body: deleteParams || {},
        });
        return result;
    }
    /**
     * Get all device associated with bucket
     * @param bucketID Bucket ID
     */
    getDevicesAssociated(bucketID) {
        const result = this.doRequest({
            path: `/bucket/${bucketID}/device`,
            method: "GET",
        });
        return result;
    }
    /**
     * Export Data from Bucket
     * @param buckets Array of JSON with get details
     * @param output Type of output
     * @param options Options of request
     */
    exportData(buckets, output, options) {
        const result = this.doRequest({
            path: `/data/export?output=${output}`,
            method: "POST",
            body: Object.assign({ buckets }, options),
        });
        return result;
    }
}
exports.default = Buckets;
