// @tago-io/sdk/out/modules/Account/Notifications.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Notifications extends TagoIOModule_1.default {
    async list(query) {
        const result = await this.doRequest({
            path: "/notification/",
            method: "GET",
            params: query,
        });
        return result;
    }
    async markAsRead(notificationIDS) {
        const result = await this.doRequest({
            path: "/notification/read",
            method: "PUT",
            body: {
                notification_ids: notificationIDS,
            },
        });
        return result;
    }
    async accept(notificationID) {
        const result = await this.doRequest({
            path: `/notification/accept/${notificationID}`,
            method: "POST",
        });
        return result;
    }
    async refuse(notificationID) {
        const result = await this.doRequest({
            path: `/notification/refuse/${notificationID}`,
            method: "POST",
        });
        return result;
    }
    async remove(notificationID) {
        const result = await this.doRequest({
            path: `/notification/${notificationID}`,
            method: "DELETE",
        });
        return result;
    }
    async registerDevice(deviceToken, platform) {
        const result = await this.doRequest({
            path: `/notification/push/register`,
            method: "POST",
            body: {
                device_token: deviceToken,
                platform,
            },
        });
        return result;
    }
    async unRegisterDevice(deviceToken) {
        const result = await this.doRequest({
            path: `/notification/push/unregister`,
            method: "POST",
            body: {
                device_token: deviceToken,
            },
        });
        return result;
    }
}
exports.default = Notifications;
