// @tago-io/sdk/out/modules/Account/PaymentMethods.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class PaymentMethods extends TagoIOModule_1.default {
    async create(paymentMethodData) {
        const result = await this.doRequest({
            path: "/account/payment_method/",
            method: "POST",
            body: paymentMethodData,
        });
        return result;
    }
    async list() {
        const result = await this.doRequest({
            path: "/account/payment_method/",
            method: "GET",
        });
        return result;
    }
    async setDefault(paymentMethodID) {
        const result = await this.doRequest({
            path: "/account/payment_method/",
            method: "PUT",
            body: {
                id: paymentMethodID,
            },
        });
        return result;
    }
    async delete(paymentMethodID) {
        const result = await this.doRequest({
            path: "/account/payment_method/",
            method: "DELETE",
            body: {
                id: paymentMethodID,
            },
        });
        return result;
    }
}
exports.default = PaymentMethods;
