// @tago-io/sdk/out/modules/Account/Plan.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
class Plan extends TagoIOModule_1.default {
    // TODO Internal Error
    async setPlanParameters(data) {
        const result = await this.doRequest({
            path: "/account/plan/",
            method: "POST",
        });
        return result;
    }
    async getActivePlan() {
        const result = await this.doRequest({
            path: "/account/plan",
            method: "GET",
        });
        return result;
    }
    async summary() {
        const result = await this.doRequest({
            path: "/billing",
            method: "GET",
        });
        return result;
    }
    async getCurrentPrices() {
        const result = await this.doRequest({
            path: "/pricing",
            method: "GET",
        });
        return result;
    }
}
exports.default = Plan;
