// @tago-io/sdk/out/common/TagoIOModule.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const qs_1 = __importDefault(require("qs"));
const apiRequest_1 = __importDefault(require("../infrastructure/apiRequest"));
const regions_1 = __importDefault(require("../regions"));
/**
 * Create a Object for Axios
 * @internal
 * @param uri URI
 * @param requestObj doRequestParams
 */
function mountAxiosRequest(uri, requestObj) {
    const axiosObj = Object.assign({ url: `${uri}${requestObj.path}`, method: requestObj.method, data: requestObj.body, params: requestObj.params, paramsSerializer: (p) => qs_1.default.stringify(p), headers: Object.assign({}, requestObj.headers) }, requestObj.overwriteAxiosConfig);
    return axiosObj;
}
/**
 * Abstract class to wrap all TagoIO SDK Modules
 * @internal
 */
class TagoIOModule {
    constructor(params) {
        this.params = {};
        this.params = params;
        this.validateParams();
    }
    validateParams() {
        if (!this.params) {
            throw new Error("Invalid Params");
        }
        if (!this.params.token) {
            throw new Error("Invalid Token");
        }
        // if (this.params.options && typeof this.params.options !== "object") {
        //   throw new Error("Invalid Params");
        // }
    }
    doRequest(requestObj) {
        var _a;
        const apiURI = (_a = regions_1.default(this.params.region)) === null || _a === void 0 ? void 0 : _a.api;
        if (!apiURI) {
            throw new Error("Invalid region");
        }
        const axiosObj = mountAxiosRequest(apiURI, requestObj);
        axiosObj.headers = Object.assign({ token: this.params.token }, axiosObj.headers);
        return apiRequest_1.default(axiosObj);
    }
    static doRequestAnonymous(requestObj, region) {
        var _a;
        const apiURI = (_a = regions_1.default(region)) === null || _a === void 0 ? void 0 : _a.api;
        if (!apiURI) {
            throw new Error("Invalid region");
        }
        const axiosObj = mountAxiosRequest(apiURI, requestObj);
        return apiRequest_1.default(axiosObj);
    }
}
exports.default = TagoIOModule;
